rainier-core/src/main/scala/com/stripe/rainier/core/Generator.scala

```diff
@@ -6,33 +6,53 @@ import com.stripe.rainier.sampler.RNG
 /**
   * Generator trait, for posterior predictive distributions to be forwards sampled during sampling
   */
-trait Generator[T] { self =>
+sealed trait Generator[T] { self =>
+  import Generator.{Const, From}
+
   def requirements: Set[Real]
 
   def get(implicit r: RNG, n: Numeric[Real]): T
 
-  def map[U](fn: T => U): Generator[U] = new Generator[U] {
-    val requirements: Set[Real] = self.requirements
-    def get(implicit r: RNG, n: Numeric[Real]): U = fn(self.get)
+  def map[U](fn: T => U): Generator[U] = self match {
+    case Const(reqs, t)     => Const(reqs, fn(t))
+    case From(reqs, fromFn) => From(reqs, (r, n) => fn(fromFn(r, n)))
   }
 
-  def flatMap[U](fn: T => Generator[U]): Generator[U] = new Generator[U] {
-    val requirements: Set[Real] = self.requirements
-    def get(implicit r: RNG, n: Numeric[Real]): U = fn(self.get).get
+  def flatMap[U](fn: T => Generator[U]): Generator[U] = self match {
+    case Const(reqsL, t) =>
+      fn(t) match {
+        case Const(reqsR, u)     => Const(reqsL ++ reqsR, u)
```

Comment:
This would be an actual behavior change, so let me comment. In the old implementation, only `self.requirements` propagated to the returned instance, NOT the internal requirements. If we're flatmapping two `Constant[T]` instances, however, we have access to both `reqL` and `reqR` and can take their union.

This isn't possible in the case where `self` is of type `From[T]`, so I left the behavior the same as old, and just passed `self.requirements` along.

Do we want to stay consistent, here? I don't know how requirements is used in the codebase, so it's up to y'all.